### PR TITLE
Backup and restore snapshot lister and delete secrets

### PR DESCRIPTION
**Pull request by ashish-amarnath**
Signed-off-by: Ashish Amarnath <ashisham@vmware.com>

Branch for this PR was cut from that of #31.

Changes in this PR:
- add BackupItemAction for VolumeSnapshotContent that:
    -  adds deletion secret, if it exists, as additional item to back up
- modify RestoreItemAction for VolumeSnapshotContent to:
    -  not modify the VolumeSnapshotContent spec
    - add deletion secret, if it exists, as additional item to restore
- modify BackupItemAction for VolumeSnapshotClass to:
    - adds snapshot lister secret, if it exists, as an additional item to backup
- modify RestoreItemAction for VolumeSnapshotClass to:
    - not modify the VolumesnapshotClass spec
    - add snapshot lister secret, if it exists as an additional item to restore

### CODE REVIEW

**Comment by ashish-amarnath** [comment]
This branch was cut from https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/31.

I know we discussed keeping this and another change in one PR. But I think it would be better to break that up into different PRs. This is a pretty straightforward change and I don't want to make this PR complicated.

File: internal/backup/volumesnapshotcontent_action.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+// AppliesTo returns information indicating that the VolumeSnapshotContentBackupItemAction action should be invoked to backup volumesnapshotcontents.
+func (p *VolumeSnapshotContentBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshotcontent.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+// Execute backs up only those VolumeSnapshotContents that are not bound with any volumesnapshot object.
```
Comment (skriss) [inline internal/backup/volumesnapshotcontent_action.go:47]: godoc can be updated

File: internal/backup/volumesnapshotcontent_action.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	var snapCont snapshotv1beta1api.VolumeSnapshotContent
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &snapCont); err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	// we should backup the snapshot deletion secrets that may be referenced in the volumesnapshotcontent's annotation
+	if util.IsVolumeSnapshotContentHasDeleteSecret(&snapCont) {
+		deleteSnapshotSecretName := snapCont.Annotations[util.PrefixedSnapshotterSecretNameKey]
```
Comment (skriss) [inline internal/backup/volumesnapshotcontent_action.go:60]: don't really need to declare these vars, can just access the values via the `snapCont.Annotations[]` map on L65-66

File: internal/backup/volumesnapshotcontent_action.go (hunk 1 above)
Comment (ashish-amarnath) [reply internal/backup/volumesnapshotcontent_action.go]: done

File: main.go [hunk 3]
```diff
@@ -30,6 +30,7 @@ func main() {
 		RegisterBackupItemAction("velero.io/csi-snapshotter", newCSISnapshotter).
 		RegisterBackupItemAction("velero.io/volumesnapshot-backupper", newVolumeSnapshotBackupItemAction).
 		RegisterBackupItemAction("velero.io/volumesnapshotclass-backupper", newVolumesnapshotClassBackupItemAction).
+		RegisterBackupItemAction("velero.io/volumesnapshotcontent-backupper", newVolumeSnapContentBackupItemAction).
 		RegisterRestoreItemAction("velero.io/csi-restorer", newCSIRestorer).
 		RegisterRestoreItemAction("velero.io/volumesnapshotclass-restorer", newVolSnapClassRestorer).
 		RegisterRestoreItemAction("velero.io/volumesnapshotcontents-restorer", newVSCRestorer).
```
Comment (skriss) [inline main.go:36]: one tiny nit: I'd make the plugin name here `velero.io/volumesnapshotcontent-restorer` (remove the `s` after `volumesnapshotcontent`) to be consistent.

Comment (ashish-amarnath) [reply main.go]: fixed!

File: internal/backup/volumesnapshotcontent_action.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		return nil, nil, errors.WithStack(err)
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	// we should backup the snapshot deletion secrets that may be referenced in the volumesnapshotcontent's annotation
+	if util.IsVolumeSnapshotContentHasDeleteSecret(&snapCont) {
+		// TODO: add GroupResource for secret into kuberesource
+		additionalItems = append(additionalItems, velero.ResourceIdentifier{
+			GroupResource: schema.GroupResource{Group: "", Resource: "secrets"},
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply internal/backup/volumesnapshotcontent_action.go:63]: This isn't available in the 1.3.2 release https://pkg.go.dev/github.com/vmware-tanzu/velero@v1.3.2/pkg/kuberesource?tab=doc

Will update once this is available in a release

File: internal/restore/volumesnapshotclass_restorer.go [hunk 5]
```diff
@@ -50,7 +52,13 @@ func (p *VolumeSnapshotClassRestoreItemAction) Execute(input *velero.RestoreItem
 	}
 
 	additionalItems := []velero.ResourceIdentifier{}
-	// TODO: place holder for returning snapshot lister secret as additional items
+	if util.IsVolumeSnapshotClassHasListerSecret(&snapClass) {
+		additionalItems = append(additionalItems, velero.ResourceIdentifier{
+			GroupResource: schema.GroupResource{Group: "", Resource: "secrets"},
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply internal/restore/volumesnapshotclass_restorer.go:57]: This isn't available in the 1.3.2 release https://pkg.go.dev/github.com/vmware-tanzu/velero@v1.3.2/pkg/kuberesource?tab=doc

Will update once this is available in a release

File: internal/restore/volumesnapshotcontents_restorer.go [hunk 6]
```diff
@@ -40,46 +41,30 @@ func (p *VSCRestorer) AppliesTo() (velero.ResourceSelector, error) {
...
-		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.ItemFromBackup from unstructured")
-	}
-
-	if vscFromBackup.Status == nil || vscFromBackup.Status.SnapshotHandle == nil {
-		return &velero.RestoreItemActionExecuteOutput{}, errors.Errorf("unable to lookup snapshotHandle from status")
+	additionalItems := []velero.ResourceIdentifier{}
+	if util.IsVolumeSnapshotContentHasDeleteSecret(&snapCont) {
+		additionalItems = append(additionalItems,
+			velero.ResourceIdentifier{
+				GroupResource: schema.GroupResource{Group: "", Resource: "secrets"},
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply internal/restore/volumesnapshotcontents_restorer.go:58]: This isn't available in the 1.3.2 release https://pkg.go.dev/github.com/vmware-tanzu/velero@v1.3.2/pkg/kuberesource?tab=doc

Will update once this is available in a release

File: internal/backup/volumesnapshotclass_action.go [hunk 7]
```diff
@@ -46,17 +48,20 @@ func (p *VolumeSnapshotClassBackupItemAction) AppliesTo() (velero.ResourceSelect
...
 	}
 
 	additionalItems := []velero.ResourceIdentifier{}
-
-	p.Log.Infof("Returning %d additionalItems to backup", len(additionalItems))
-	for _, ai := range additionalItems {
-		p.Log.Debugf("%s: %s", ai.GroupResource.String(), ai.Name)
+	if util.IsVolumeSnapshotClassHasListerSecret(&snapClass) {
+		additionalItems = append(additionalItems, velero.ResourceIdentifier{
+			GroupResource: schema.GroupResource{Group: "", Resource: "secrets"},
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply internal/backup/volumesnapshotclass_action.go:59]: This isn't available in the 1.3.2 release https://pkg.go.dev/github.com/vmware-tanzu/velero@v1.3.2/pkg/kuberesource?tab=doc

Will update once this is available in a release

File: internal/backup/volumesnapshotcontent_action.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+// VolumeSnapshotContentBackupItemAction is a backup item action plugin to backup
+// CSI VolumeSnapshotcontent objects using Velero
+type VolumeSnapshotContentBackupItemAction struct {
+	Log logrus.FieldLogger
+}
+
+// AppliesTo returns information indicating that the VolumeSnapshotContentBackupItemAction action should be invoked to backup volumesnapshotcontents.
+func (p *VolumeSnapshotContentBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply internal/backup/volumesnapshotcontent_action.go]: done

**Pull request merged**